langkit/ocaml_api.py

```diff
@@ -16,19 +16,46 @@ class AnalysisContextType(object):
         Placeholder to represent the analysis context type in our depedency
         tracking system (see the type_graph and add_dep methods below).
         """
-        pass
+        is_entity_type = False
+        is_array_type = False
+        dsl_name = 'AnalysisContext'
+
+    def actual_c_type(self, typ):
+        """
+        Return the C type used to encode ``typ`` values.
+
+        In the C/OCaml binding layer, some types are encoded as other types.
+        For instance: all entities are exposed as the root entity type.
+
+        :param CompiledType typ: Type to encode in the C/OCaml binding layer.
+        :rtype: CompiledType
+        """
+        if typ.is_entity_type:
+            # Expose all entity types as the root entity type
+            result = T.entity
+        elif typ.is_array_type:
+            result = self.actual_c_type(typ.element_type).array
+        else:
+            result = typ
+
+        if not isinstance(result, self.AnalysisContextType):
+            result = resolve_type(result)
```

Comment:
Why do you have a special case on `self.AnalysisContextType` ? This code is not very clear, and anyway `resolve_type` is made to sort out that kind of stuff by itself, why have a condition ?

Reply:
AnalysisContextType is not a CompiledType (see above in this module): it’s a hack in the ocaml bindings generator to properly order type declarations, including analysis contexts. So resolve_type does not work for it.

Reply:
Ok it probably deserves a comment then!

Reply:
Fair enough. Comment added, thanks.